eventuals/http.h

```diff
@@ -645,48 +767,66 @@ struct _HTTP {
 
 ////////////////////////////////////////////////////////////////////////
 
-inline auto Get(
-    EventLoop& loop,
-    const std::string& url,
-    const std::chrono::nanoseconds& timeout = std::chrono::nanoseconds(0)) {
+inline auto Client::Do(Request&& request) {
+  // TODO(benh): need 'Client::Default()'.
+  auto& loop = EventLoop::Default();
+
   // NOTE: we use a 'RescheduleAfter()' to ensure we use current
   // scheduling context to invoke the continuation after the transfer has
   // completed (or was interrupted).
   return RescheduleAfter(
-      // TODO(benh): borrow 'this' so http call can't outlive a loop.
-      _HTTP::Composable{loop, url, Method::GET, timeout});
+      // TODO(benh): borrow '&loop' so http call can't outlive a loop.
+      _HTTP::Composable{
+          loop,
+          request.uri(),
+          request.method(),
+          request.timeout(),
+          request.fields()});
 }
 
 ////////////////////////////////////////////////////////////////////////
 
-inline auto Get(
-    const std::string& url,
-    const std::chrono::nanoseconds& timeout = std::chrono::nanoseconds(0)) {
-  return Get(EventLoop::Default(), url, timeout);
+inline auto Client::Get(
+    std::string&& uri,
+    std::chrono::nanoseconds&& timeout) {
+  return Do(
+      Request::Builder()
+          .uri(std::move(uri))
+          .method(GET)
+          .timeout(std::move(timeout))
+          .Build());
 }
 
 ////////////////////////////////////////////////////////////////////////
 
-inline auto Post(
-    EventLoop& loop,
-    const std::string& url,
-    const PostFields& fields,
-    const std::chrono::nanoseconds& timeout = std::chrono::nanoseconds(0)) {
-  // NOTE: we use a 'RescheduleAfter()' to ensure we use current
-  // scheduling context to invoke the continuation after the transfer has
-  // completed (or was interrupted).
-  return RescheduleAfter(
-      // TODO(benh): borrow '&loop' so http call can't outlive a loop.
-      _HTTP::Composable{loop, url, Method::POST, timeout, fields});
+inline auto Client::Post(
+    std::string&& uri,
+    PostFields&& fields,
+    std::chrono::nanoseconds&& timeout) {
+  return Do(
+      Request::Builder()
+          .uri(std::move(uri))
+          .method(POST)
+          .timeout(std::move(timeout))
+          .fields(std::move(fields))
+          .Build());
+}
+
+////////////////////////////////////////////////////////////////////////
+
+inline auto Get(
+    std::string&& url,
+    std::chrono::nanoseconds&& timeout = std::chrono::nanoseconds(0)) {
```

Comment:
I do see it working with 
`Get("http://someurl.com", std::chrono::seconds(10))`, 
but it will not work with
```
const std::string hostname = "http://someurl.com";
const auto timeout = std::chrono::seconds(10);
Get(hostname, timeout);
```